### PR TITLE
Add the general convolution operation to extensions

### PR DESCRIPTION
**Pull request by DarrenZhang01**
TF XLA version:
https://www.tensorflow.org/xla/operation_semantics?hl=en#conv_convolution;
JAX version: https://jax.readthedocs.io/en/stable/_autosummary/jax.lax.conv_general_dilated.html

### CODE REVIEW

**Comment by DarrenZhang01** [comment]
tests coming soon.

**Comment by akshaym** [comment]
Hey @DarrenZhang01, let me know if I can take a look at this. If there are bugs lets just document them for now and add the code anyway? WDYT?

**Comment by DarrenZhang01** [comment]
> Hey @DarrenZhang01, let me know if I can take a look at this. If there are bugs lets just document them for now and add the code anyway? WDYT?

Thanks very much for the support, Akshay! It is ready for being reviewed.

File: trax/tf_numpy/extensions/extensions.py [hunk 1]
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
 
+# TODO (Zhibo Zhang): Run pylint and complement the docstring.
+def _conv_general_param_type_converter(window_strides, lhs_dilation, rhs_dilation):
+  """ Convert the inputs strides, lhs_dilation, rhs_dilation to the standard
+  TF conv inputs.
+  For example,
+   in the 3D case, if lhs_dilation = 2, then convert it to [2, 2, 2]
+                   if lhs_dilation = (2, 2, 2), convert it also to [2, 2, 2]
+  """
+  strides = [window_strides] * dim if isinstance(window_strides, int) else \
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:590]: Where is dim coming from? 

Alternatively you can define an inner function and just use it on each:
```suggestion
def _as_list_of_size(item, size):
  if item is None:
    return None
  return [item] * size if isinstance(item, int) else list(item)
  
  return _as_list_of_size(window_strides), _as_list_of_size(lhs_dilation), _as_list_of_size(rhs_dilation)
```

File: trax/tf_numpy/extensions/extensions.py [hunk 2]
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+#                     the according bugs.
+# TODO (Zhibo Zhang): Support feature_group_count, batch_group_count and precision, and
+#       allow lhs_dilation and rhs_dilation to happen at the same time.
+def conv_general_dilated(lhs, rhs, window_strides, padding, output_shape,
+                         lhs_dilation=None, rhs_dilation=None,
+                         dimension_numbers=None, feature_group_count=1,
+                         batch_group_count=1, precision=None):
+  """ A general conv API that integrates normal conv, deconvolution,
+  dilated convolution, etc."""
+  dim = None
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:612]: nit: can remove this line?

File: trax/tf_numpy/extensions/extensions.py [hunk 3]
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+  lhs_spec, rhs_spec, out_spec = dimension_numbers
+  if lhs_spec != out_spec:
+    raise TypeError("Current implementation requires the `data_format` of the "
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:615]: These and the below all feel like ValueErrors to me - not TypeErrors.

File: trax/tf_numpy/extensions/extensions.py [hunk 4]
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+    dim_maps['O'] = rhs_spec[0]
+    dim_maps['N'] = lhs_spec[0]
+    dim_maps['C'] = lhs_spec[1]
+
+  lhs = np.moveaxis(lhs, (dim_maps['N'], dim_maps['C']), (0, dim + 1))
+  # Adjust the filters, put the dimension 'I' and 'O' at last.
+  rhs = np.moveaxis(rhs, (dim_maps['O'], dim_maps['I']), (dim + 1, dim))
+  spatial_dim_maps = {1: 'W', 2: "HW", 3: "DHW"}
+  data_format = 'N' + spatial_dim_maps[dim] + 'C'
+  tf_nn_APIs = {1: [nn.conv1d, nn.conv1d_transpose],
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:654]: Lets make this a module-level constant instead?

File: trax/tf_numpy/extensions/extensions.py [hunk 5]
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+                3: [nn.conv3d, nn.conv3d_transpose]}
+
+  output = None
+  if rhs_dilation or (lhs_dilation is None and rhs_dilation is None):
+    output = tf_nn_APIs[dim][0](lhs, rhs, strides, padding, data_format,
+                                rhs_dilation)
+  else:
+    output = tf_nn_APIs[dim][1](lhs, rhs, tf.constant(output_shape), strides,
+                                padding, data_format, lhs_dilation)
+  output = np.moveaxis(output, (0, dim + 1), (dim_maps['N'], dim_maps['C']))
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:665]: I think this and below should be using tf_np.moveaxis and tf_np.asarray. 

It'd be good to document the input output expectations. Perhaps similar to dot_general, the name for this function should be "tf_conv_general_dilated", so inputs and outputs are expected to be tf.Tensors, and the tf_np.asarray should not be required.

File: trax/tf_numpy/extensions/extensions_test.py [hunk 6]
```diff
@@ -421,6 +422,53 @@ def test_tf_dot_general(self, lhs_np, rhs_np, dims):
...
+                             feature_group_count, batch_group_count,
+                             dimension_numbers, perms):
+    tf.print("dimension_numbers: {}".format(dimension_numbers), output_stream=sys.stdout)
+    lhs_perm, rhs_perm = perms  # permute to compatible shapes
+
+    lhs_tf = tf_np.transpose(tf_np.ones(lhs_shape), lhs_perm)
+    rhs_tf = tf_np.transpose(tf_np.ones(rhs_shape), rhs_perm)
+
+    lhs_jax = jnp.transpose(jnp.ones(lhs_shape), lhs_perm)
+    rhs_jax = jnp.transpose(jnp.ones(rhs_shape), rhs_perm)
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions_test.py:461]: Instead of these, don't conv_general_dilated from JAX and TF both accept numpy arrays?

File: trax/tf_numpy/extensions/extensions_test.py [hunk 7]
```diff
@@ -421,6 +422,53 @@ def test_tf_dot_general(self, lhs_np, rhs_np, dims):
...
+        (None, (1, 1))
+      ]
+      for dimension_numbers, perms in [
+        (("NHWC", "HWIO", "NHWC"), ([0, 2, 3, 1], [2, 3, 1, 0]))
+      ]])
+  def testConvGeneralDilated(self, lhs_shape, rhs_shape, strides,
+                             padding, lhs_dilation, rhs_dilation,
+                             feature_group_count, batch_group_count,
+                             dimension_numbers, perms):
+    tf.print("dimension_numbers: {}".format(dimension_numbers), output_stream=sys.stdout)
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions_test.py:454]: Remove debug print?

File: trax/tf_numpy/extensions/extensions_test.py [hunk 8]
```diff
@@ -421,6 +422,53 @@ def test_tf_dot_general(self, lhs_np, rhs_np, dims):
...
+    lhs_tf = tf_np.transpose(tf_np.ones(lhs_shape), lhs_perm)
+    rhs_tf = tf_np.transpose(tf_np.ones(rhs_shape), rhs_perm)
+
+    lhs_jax = jnp.transpose(jnp.ones(lhs_shape), lhs_perm)
+    rhs_jax = jnp.transpose(jnp.ones(rhs_shape), rhs_perm)
+
+    jax_conv = jax.lax.conv_general_dilated(lhs_jax, rhs_jax, strides, padding, lhs_dilation,
+      rhs_dilation, dimension_numbers, feature_group_count, batch_group_count)
+
+    tf_conv = lax.conv_general_dilated(lhs_tf, rhs_tf, strides, padding, jax_conv.shape, lhs_dilation,
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions_test.py:466]: extensions.conv_general_dilated?

File: trax/tf_numpy/extensions/extensions.py [hunk 9]
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+    lhs_dilation = [lhs_dilation] * dim if isinstance(lhs_dilation, int) else \
+                    list(lhs_dilation)
+  if rhs_dilation:
+    rhs_dilation = [rhs_dilation] * dim if isinstance(rhs_dilation, int) else \
+                    list(rhs_dilation)
+  return (strides, lhs_dilation, rhs_dilation)
+
+
+# TODO (Zhibo Zhang): Run pylint and complement the docstring.
+# TOTO (Zhibo Zhang): Expand the test cases of general convolution and revise
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:602]: nit spelling TOTO -> TODO

File: trax/tf_numpy/extensions/extensions.py [hunk 10]
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+  lhs = np.moveaxis(lhs, (dim_maps['N'], dim_maps['C']), (0, dim + 1))
+  # Adjust the filters, put the dimension 'I' and 'O' at last.
+  rhs = np.moveaxis(rhs, (dim_maps['O'], dim_maps['I']), (dim + 1, dim))
+  spatial_dim_maps = {1: 'W', 2: "HW", 3: "DHW"}
+  data_format = 'N' + spatial_dim_maps[dim] + 'C'
+  tf_nn_APIs = {1: [nn.conv1d, nn.conv1d_transpose],
+                2: [nn.conv2d, nn.conv2d_transpose],
+                3: [nn.conv3d, nn.conv3d_transpose]}
+
+  output = None
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:658]: nit: can remove this line?

File: trax/tf_numpy/extensions/extensions.py [hunk 11]
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
   return tf.einsum(equation, lhs, rhs)
 
 
+# TODO (Zhibo Zhang): Run pylint and complement the docstring.
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:568]: Lets do both pylint/docstring changes for both functions this and below before submission. 

Also, I think its better to put the TODOs under your github username instead of your full name (so "DarrenZhang01" instead of "Zhibo Zhang").

**Comment by DarrenZhang01** [comment]
> You had mentioned a bug - is that still around? Or is that fixed?

So the original bug is that the transpose convolution was not supported due to the lack of output shape as I mentioned in our last Friday's meeting. Now I added an output shape evaluation helper. 

Besides, I used the test cases from JAX general convolution. Those cases are really broad and TF general conv currently do not cover them all, so I added a TODO to expand the test cases later on.

File: trax/tf_numpy/extensions/extensions.py (hunk 4 above)
Comment (DarrenZhang01) [reply trax/tf_numpy/extensions/extensions.py:657]: Hi, Akshay, could you elaborate on how to make this "module-level constant"?

Comment (DarrenZhang01) [reply trax/tf_numpy/extensions/extensions.py:657]: OK, I will do something similar to `_int_dtypes`.

File: trax/tf_numpy/extensions/extensions_test.py (hunk 6 above)
Comment (DarrenZhang01) [reply trax/tf_numpy/extensions/extensions_test.py:461]: Yeah, I think vanilla NumPy arrays should not affect function execution here.

File: trax/tf_numpy/extensions/extensions.py (hunk 2 above)
Comment (DarrenZhang01) [reply trax/tf_numpy/extensions/extensions.py:615]: I will send this as parameters for the helper function.

**Comment by DarrenZhang01** [comment]
Thanks for the review, Akshay!

File: trax/tf_numpy/extensions/extensions.py [hunk 12]
```diff
@@ -37,6 +38,9 @@
     tf.int64, tf.int32, tf.int16, tf.int8, tf.uint8, tf.uint16, tf.uint32,
     tf.uint64
 ]
+_tf_nn_APIs = {1: [nn.conv1d, nn.conv1d_transpose],
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:41]: nit just use tf.nn.conv1d etc and remove the additional import.

File: trax/tf_numpy/extensions/extensions.py (hunk 2 above)
Comment (akshaym) [reply trax/tf_numpy/extensions/extensions.py:625]: Setting dim = None is still not required though right? Line 633 below sets dim anyway?

**Comment by DarrenZhang01** [comment]
> Thanks for the review, Akshay!

Now it is ready for review again. @akshaym

File: trax/tf_numpy/extensions/extensions.py [hunk 13]
```diff
@@ -564,6 +568,113 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+    rhs_dilation: None, or a sequence of n integers, giving the dilation factor
+                  to apply in each spatial dimension of rhs. RHS dilation is
+                  also known as atrous convolution.
+    dimension_numbers: either None, a ConvDimensionNumbers object, or a 3-tuple
+                       (lhs_spec, rhs_spec, out_spec), where each element is a
+                       string of length n+2.
+    feature_group_count:  integer, default 1.
+    batch_group_count: integer, default 1.
+    precision: Optional. Either None, which means the default precision for the
+               backend, or a Precision enum value.
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:623]: Can you add a Returns section specifying that a tf_np.ndarray is returned?

File: trax/tf_numpy/extensions/extensions.py [hunk 14]
```diff
@@ -564,6 +568,113 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+    lhs_dilation: None, or a sequence of n integers, giving the dilation factor
+                  to apply in each spatial dimension of lhs. LHS dilation is
+                  also known as transposed convolution.
+    rhs_dilation: None, or a sequence of n integers, giving the dilation factor
+                  to apply in each spatial dimension of rhs. RHS dilation is
+                  also known as atrous convolution.
+    dimension_numbers: either None, a ConvDimensionNumbers object, or a 3-tuple
+                       (lhs_spec, rhs_spec, out_spec), where each element is a
+                       string of length n+2.
+    feature_group_count:  integer, default 1.
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:620]: Add comments for things that are currently not supported. 
For example:
```suggestion
    feature_group_count:  integer, default 1. Changing this is currently not supported. 
```

File: trax/tf_numpy/extensions/extensions.py [hunk 15]
```diff
@@ -564,6 +568,113 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+    if lhs_dilation == (1,) * dim and rhs_dilation == (1,) * dim:
+      lhs_dilation, rhs_dilation = None, None
+    else:
+      raise ValueError("Current implementation does not support that "
+                       "deconvolution and dilation to be performed at the same "
+                       "time, but got lhs_dilation: {}, rhs_dilation: {}"
+                       .format(lhs_dilation, rhs_dilation))
+  if padding not in ["SAME", "VALID"]:
+    raise ValueError("Current implementation requires the padding parameter"
+                     "to be either 'VALID' or 'SAME', but got: ", padding)
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:644]: Lets also check that feature_group_count, batch_group_count and precision are equal to their defaults. If they are not, we should raise a NotImplementedError.

File: trax/tf_numpy/extensions/extensions_test.py [hunk 16]
```diff
@@ -421,6 +424,56 @@ def test_tf_dot_general(self, lhs_np, rhs_np, dims):
...
+  def testConvGeneralDilated(self, lhs_shape, rhs_shape, strides,
+                             padding, lhs_dilation, rhs_dilation,
+                             feature_group_count, batch_group_count,
+                             dimension_numbers, perms):
+    lhs_perm, rhs_perm = perms  # permute to compatible shapes
+
+    lhs = np.transpose(np.ones(lhs_shape), lhs_perm)
+    rhs = np.transpose(np.ones(rhs_shape), rhs_perm)
+
+    jax_conv = jax.lax.conv_general_dilated(lhs, rhs, strides, padding,
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions_test.py:461]: lax.conv_general_dilated and remove the import jax statement from above.

File: trax/tf_numpy/extensions/extensions_test.py [hunk 17]
```diff
@@ -421,6 +424,56 @@ def test_tf_dot_general(self, lhs_np, rhs_np, dims):
...
+                                            lhs_dilation, rhs_dilation,
+                                            dimension_numbers,
+                                            feature_group_count,
+                                            batch_group_count)
+
+    tf_conv = extensions.tf_conv_general_dilated(lhs, rhs, strides,
+                                                 padding, jax_conv.shape,
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions_test.py:468]: Can you make this just be a None instead of jax_conv.shape? Looks like we are not testing lhs_dilation in any case?

File: trax/tf_numpy/extensions/extensions.py [hunk 18]
```diff
@@ -564,6 +568,113 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+
+  Args: (Use JAX documentation as a reference)
+    lhs: a rank n+2 dimensional input array.
+    rhs: a rank n+2 dimensional array of kernel weights.
+    window_strides: a sequence of n integers, representing the inter-window
+                    strides.
+    padding: either the string ‘SAME’, the string ‘VALID’, or a sequence of n
+             (low, high) integer pairs that give the padding to apply before and
+             after each spatial dimension.
+    output_shape: the output shape of the convolution.
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:610]: Point out that its only required for transposed convolution?

**Comment by DarrenZhang01** [comment]
Thanks a lot, Akshay!

File: trax/tf_numpy/extensions/extensions.py [hunk 19]
```diff
@@ -564,6 +567,125 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
...
+                       .format(lhs_dilation, rhs_dilation))
+  if padding not in ["SAME", "VALID"]:
+    raise ValueError("Current implementation requires the padding parameter"
+                     "to be either 'VALID' or 'SAME', but got: ", padding)
+  if batch_group_count != 1 or feature_group_count != 1:
+    raise NotImplementedError("batch_group_count and feature_group_count "
+                              "other than 1 is currently not supported, but"
+                              " got feature_group_count: {}, batch_group_count"
+                              ": {}".format(feature_group_count,
+                                            batch_group_count))
```
Comment (akshaym) [inline trax/tf_numpy/extensions/extensions.py:655]: Check precision also?
